### ISSUE TITLE
Cannot build a tag with several arguments and applied filters

### ISSUE THREAD

**Issue by ouven** [markglenn/liquex#24]
Hi @markglenn 

we are trying to bild a tag, that can render an other component, like:
```liquid
{% render "component_name", binding1: 'foo' %}
```
This is easy to achieve, thanks to you cool extension mechanism.

## problem
But now we wanted the component to be an argument with filters, which cannot be achieved:
```liquid
{% render "component_name" | append: a_variables_content,
    binding1: 'foo' | bar: buzz, 'foo2',
    binding2: 'foo3'
%}
```
The problem is, that the argument list of a filter can have a keyword list at the end, which makes a string with a colon at the end ambigues (lib/liquex/parser/object.ex:36).

What is this keyword list for filters needed for? I haven't found a test for it nor have I found a shopify liquid example like that?

I'm asking for curiousity. I'm Still a big fan of this project :)

Cheers

**Comment by markglenn** [comment]
Thank you for pointing this out. I honestly don't remember the reason behind allowing a keyword list at the end of a filter. I will have to go back and look at some of the examples I worked with in the past and my current custom component implementations and how they work with it.

I will look into it to see how the Liquid gem handles your use case since it really is ambiguously defined in the documents and test cases.

Just for clarity, how would you expect your render tag to be parsed? I know it's not standard liquid syntax, but where would you place parentheses around your block of code to make it not ambiguous? Where are you expecting the filters to be applied?

**Comment by ouven** [comment]
Hi @markglenn 

I thought the ambiguity, would have been gone, when the keyword list is not parsed, because of the colon, that would mark the next binding of a render, like:

* filter is marked with a pipe symbol: `(| [^\s:]+:?( [^, ]+,?)*)*`
* binding parameter is marked without a pipe symbol: `([^\s:]+:? [^,\s]+,?)*`

Regexes are not exact, but I hope they show my idea.

**Comment by markglenn** [comment]
Hi @ouven,

I'm coming back to this issue and I'm again trying to dig into the issue.

I found where keyword lists at the end of a filter come into play. They don't exist in the base liquid gem, but do show up in the Shopify liquid documentation under the [image_url](https://shopify.dev/api/liquid/filters/url-filters#image_url) filter.  In their example, they allow a keyword list to add URL parameters to an image:

    {{ product | image_url: width: 100, height: 100 }}

You can find the test that showcases that in the code base at [parser/object_test.exs:33](https://github.com/markglenn/liquex/blob/master/test/liquex/parser/object_test.exs#L33). This test is pulled directly from my use case in Enology for Winestyr where this library is used.

In your use case, Liquid should parse this as [parentheses added for clarity]:

    {% render "component_name" |
        (append: a_variables_content, binding1: 'foo') |
        (bar: buzz, 'foo2', binding2: 'foo3') %}

Or in plain English:

1. Render out the render custom tag
2. Run that result through the "append" filter with arguments `a_variables_content` and `binding1: 'foo'`
3. Run that result through the "bar" filter with arguments `buzz`, `'foo2'`, and `binding2: 'foo3'`

And that seems to follow what Shopify describes in their Liquid for Shopify documentation.

If I'm missing something, maybe that's because you're trying to write it in a generic way here with 'foo', 'buzz', etc. Can you give me a more concrete example so we can maybe find a work around or to help me understand how this library is not following the standard?

**Comment by ouven** [comment]
Hi @markglenn,

What we were trying to achieve was more like the following scenario:
1. we have a custom `render` tag, that can render another template with isolated context bindings bar: "buzz".
2. the name of the template is dynamic and is built by adding `a_variables_content` to our `component_name_prefix`.

```liquid
{% render ("component_name_prefix" |  append: a_variables_content), bar: buzz %}
```

Now, by the fact, that a filter can have a keyword list, I cannot determine where the keyword list of the filter end and the binding list of the render tag starts. We can work around this easily, by just doing something like:

```liquid
{% assign component_name = "component_name_prefix" |  append: a_variables_content %}
{% render component_name, bar: buzz %}
```

Actually I was not aware of the use case, that you did mention. I did not find an example for a dynamic variable list for a filter. I was only looking at the more generic documentation: https://shopify.github.io/liquid/filters

So I guess, we cannot solve this ambiguity and just keep assignments as replacement for parentheses :)
